crates/storage/src/connection/trie.rs

```diff
@@ -222,33 +250,85 @@ impl Transaction<'_> {
         self.trie_node_hash(index, "trie_storage")
     }
 
-    fn delete_contract_roots(&self, contract: ContractAddress) -> anyhow::Result<()> {
-        let mut stmt = self
-            .inner()
-            .prepare_cached("DELETE FROM contract_roots WHERE contract_address = ?")?;
-        stmt.execute(params![&contract])?;
+    fn delete_contract_roots(
+        &self,
+        contract: ContractAddress,
+        before_block: BlockNumber,
+    ) -> anyhow::Result<()> {
+        let mut stmt = self.inner().prepare_cached(
+            "DELETE FROM contract_roots WHERE contract_address = ? AND block_number < ?",
+        )?;
+        stmt.execute(params![&contract, &before_block])?;
         Ok(())
     }
 
-    fn remove_trie(&self, removed: &[u64], table: &'static str) -> anyhow::Result<()> {
-        let mut stmt = self
-            .inner()
-            .prepare_cached(&format!("DELETE FROM {table} WHERE idx = ?"))
-            .context("Creating delete statement")?;
+    fn remove_trie(
+        &self,
+        removed: &[u64],
+        block_number: BlockNumber,
+        num_blocks_kept: u64,
+        table: &'static str,
+    ) -> anyhow::Result<()> {
+        if let Some(before_block) = block_number.checked_sub(num_blocks_kept) {
+            // Delete nodes that have already been marked as ready for deletion.
+            let mut select_stmt = self
+                .inner()
+                .prepare(&format!(
```

Comment:
These queries seem to be on the hot path -- should we maybe use `prepare_cached()` here?